teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,410 @@
+- `Tasks` to emit a list of artifacts of unknown length.
+
+### Requirements
+
+- Enable `Tasks` to be able to surface Artifact metadata as an array.
+
+## Proposal
+
+Define the structure of the Artifact metadata that the `Step` needs to write out. Define the file path and variable replacement syntax for the location of the metadata file.
+
+The following chapters in the proposal describe in details the proposed phased implementation approach:
+
+- [Artifact Provenance Data](#artifact-metadata)
+- [surfacing artifact metadata into the status](#from-step-into-status)
+- [surfacing the artifact metadata into the provenance payload](#from-status-into-provenance)
+
+
+### Artifact Provenance Data
+
+An artifact can be consumed or produced by a `Task`. This means that we will always have two categories at a high level: `inputs` (to indicate what was downloaded) and `outputs` (to indicate what was uploaded). The metadata content could be defined as follows:
+
+```json
+{
+    "inputs":[
+        {
+            "name": "artifact category 1", # optional, e.g. source code, dependencies # Single name for the list of `uri`, `digest` pairs.
+            "values": [
+                {
+                    "uri": string,
+                    "digest": {
+                      string: string # e.g. "sha256": "jksjd39dj39"
+                    }
+                },
+            ]
+        }
+    ],
+    "outputs": [
+        {
+            "name": "artifact category 1", # optional, e.g. images, packages, coverage reports
+            "values": [
+                {
+                    "uri": string,
+                    "digest": {
+                      string: string # e.g. "sha256": "jksjd39dj39"
+                    }
+                },
+            ]
+        }
+    ]
+}
+```
+The metadata content is written by the `Step` to a file `$(step.artifacts.path)`:
+
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee $(step.artifacts.path)
+```
+The path `$(step.artifacts.path)` resolves to `/tekton/steps/<step-name>/artifacts/provenance.json` in the `Step` container at runtime.
+```yaml
+taskSpec:
+  steps:
+    - name: artifact-reporting
+      image: bash:latest
+      script: |
+        #!/usr/bin/env bash
+        # Download some artifacts
+        # Upload some artifacts
+        echo -n "{\"inputs\":[{\"name\":\"artifact1\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}], \"outputs\":[{\"name\":\"artifact2\",\"values\":[{\"url\":\"some-uri\",\"digest\":\"some-digest\"}]}]}" | tee /tekton/steps/artifact-reporting/artifacts/provenance.json
+```
+
+#### Name
```

Comment:
Might be worth adding what constraints/validation (if any) we're applying to these fields.